app/src/main/java/ch/hikemate/app/model/elevation/ElevationServiceRepository.kt

```diff
@@ -1,113 +1,270 @@
+          launchUpdateOfCache()
+        }
+      }
     }
+  }
 
-    // Create the JSON body for the request
-    val jsonBody = buildJsonObject {
-      // The major object
-      put(
-          LOCATION_KEY,
-          buildJsonArray {
-            coordinates.forEach { coordinate ->
-              add(
-                  // Object with the location
-                  buildJsonObject {
-                    put(LATITUDE_KEY, coordinate.lat)
-                    put(LONGITUDE_KEY, coordinate.lon)
-                  })
+  /**
+   * Launch the update of the cache
+   *
+   * The update will be delayed by 1 second to allow for multiple requests to be batched together
+   *
+   * @param onSuccess The callback to be called when the request is successful
+   * @param onFailure The callback to be called when the request fails
+   */
+  private fun launchUpdateOfCache(
+      onSuccess: (List<Double>) -> Unit = {},
+      onFailure: (Exception) -> Unit = {}
+  ) {
+
+    if (requests.isEmpty()) return
+
+    // We need to start a new job to update the cache
+    cacheUpdateJob =
+        CoroutineScope(dispatcher).launch {
+          // Delay the update by some amount of time to allow for multiple requests
+          // to be batched together
+          // This is purely empirical and can be adjusted
+          // If we already have enough data for a request, we don't need to wait
+          var coordinates: List<LatLong> = emptyList()
+          mutex.withLock { coordinates = requests.flatMap { it.coordinates } }
+          if (coordinates.size < MAX_COORDINATES_PER_REQUEST) {
+            delay(CACHE_UPDATE_DELAY)
+          }
+          mutex.withLock {
+            val chunks = coordinates.chunked(MAX_COORDINATES_PER_REQUEST)
+
+            for (chunk in chunks) {
+              // Create the JSON body for the request
+              val jsonString = buildJsonArray {
+                chunk.forEach { coordinate ->
+                  add(
+                      buildJsonArray {
+                        add(coordinate.lat)
+                        add(coordinate.lon)
+                      })
+                }
+              }
+
+              val request =
+                  okhttp3.Request.Builder().url(BASE_URL + jsonString.toString()).get().build()
+
+              // Callback for the network request, it also updates the cache
+              val onSuccessWithCache: (List<Double?>) -> Unit =
+                  parseCacheOnDataRetrieval(chunk, onSuccess)
+
+              val onFailureCallback: (Exception) -> Unit = { exception ->
+                CoroutineScope(dispatcher).launch {
+                  mutex.withLock { requests.forEach { request -> request.onFailure(exception) } }
+                  onFailure(exception)
+                }
+              }
+
+              client
+                  .newCall(request)
+                  .enqueue(ElevationServiceCallback(onSuccessWithCache, onFailureCallback))
             }
```

Comment:
I'm not sure to understand what you mean? If the `onFailureCallback` is called, then it's because of an issue that retrying couldn't fix

Reply:
Since you use several chunks, each chunk gives one request to the API, right? And the requests for each individual chunk are started in parallel all at the same time, if I understand correctly.

But there is only one variable `failedRequests`. Does this mean that if I have five chunks, and each of them fails once, then I reached my limit in terms of retrying? Even though I did not retry at all?

Reply:
Ooooh, I see! Thanks! That's a flaw!

Reply:
I've rethought all of my code to change it!

Reply:
@MaelImhof Can I resolve this one?

Reply:
Got lost in all those conversations, sorry, forgot about this one 😂